### PR TITLE
Feature - global opacity

### PR DESCRIPTION
**Pull request by bbirchnz**
Global opacity control. Intended to support different opacity based on gaze, but not sure how to get gaze state back (calculated at the HMD level)

Builds on previous PR adding pen settings as relies on kneeboardstate object.

### CODE REVIEW

**Comment by fredemmott** [comment]
I'd prefer not to put this in to the tabs; this will mean it needs consistently doing for every tab type in the future, and several more here. There's also an existing non-VR overall opacity, which this conflicts with, and it would be good to keep using the system colors where appropriate.

There's two main approaches that could do that:
1. unify with non-VR overall opacity, but keep gaze opacity as VR-specific. Implement in InterprocessRenderer::CopyPixelsToSHM - replace the CopyResource (texture copy) with ID2D1DeviceContext to the shared texture - https://github.com/fredemmott/OpenKneeboard/blob/ebd5b1502591491e3d49ffd7f9a6f45fc37e9567/src/app/app-common/InterprocessRenderer.cpp#L128
2. keep as a separate option, but implement in the OpenXR/OpenVR/Oculus renderers

Given you want an option for opacity varying by gaze, (2) is most likely the way to go - right now, there is no communication from the VR implementations back to openkneeboard, so tab-based rendering can't be affected by where you're looking. That /could/ be added, but doing (2) directly (similar to 1, this would be replacing a resource/subresource copy with a paint operation, just in different places) would be simpler, and similar to how gaze-based zoom already works

**Comment by bbirchnz** [comment]
> I'd prefer not to put this in to the tabs; this will mean it needs consistently doing for every tab type in the future, and several more that already exist. There's also an existing non-VR overall opacity, which this conflicts with, and it would be good to keep using the system colors where appropriate.
> 
> There's two main approaches that could do that:
> 
> 1. unify with the existing non-VR overall opacity option. Implement in InterprocessRenderer::CopyPixelsToSHM - replace the CopyResource (texture copy) with `ID2D1DeviceContext::DrawBitmap` to the shared texture - https://github.com/fredemmott/OpenKneeboard/blob/ebd5b1502591491e3d49ffd7f9a6f45fc37e9567/src/app/app-common/InterprocessRenderer.cpp#L128
> 2. keep as a separate option, but implement in the OpenXR/OpenVR/Oculus renderers
> 
> Given you want an option for opacity varying by gaze, (2) is most likely the way to go - right now, there is no communication from the VR implementations back to openkneeboard, so tab-based rendering can't be affected by where you're looking. That /could/ be added, but doing (2) directly (similar to 1, this would be replacing a resource/subresource copy with a paint operation, just in different places) would be simpler, and similar to how gaze-based zoom already works

I like option 2 - and yeah that was the only place I could manage the gaze based stuff. Will look into it, thanks for reviewing.

File: src/app/app-common/InterprocessRenderer.cpp [hunk 1]
```diff
@@ -120,12 +122,31 @@ void InterprocessRenderer::Impl::CopyPixelsToSHM() {
     return;
   }
 
+  // redraw the whole bitmap with global opacity applied:
+  auto D2DContext = mDXR.mD2DDeviceContext;
```
Comment (fredemmott) [inline src/app/app-common/InterprocessRenderer.cpp:126]: `d2dContext` for locals

**Review by fredemmott — COMMENTED** [review]
Just copying from discord to here:

This seems good - however, once I'm done splitting out common code from the various VR renderers, I'll take a look at applying the same technique there instead. This will allow separate vr vs non-vr opacity, and separate gaze-based opacity

If that turns out not to be practical, very likely to merge this as-is

**Comment by bbirchnz** [comment]
> Just copying from discord to here:
> 
> This seems good - however, once I'm done splitting out common code from the various VR renderers, I'll take a look at applying the same technique there instead. This will allow separate vr vs non-vr opacity, and separate gaze-based opacity
> 
> If that turns out not to be practical, very likely to merge this as-is

Yep this is on hold pending outcomes of your work there.

**Comment by fredemmott** [comment]
Do you still plan to work on this PR, given the VR opacity settings that I added yesterday?

**Comment by bbirchnz** [comment]
> Do you still plan to work on this PR, given the VR opacity settings that I added yesterday?

Nope, superseded by your work.

**Pull request closed without merging**